### PR TITLE
Split host configuration and app configuration

### PR DESCRIPTION
**Pull request by davidfowl**
- Clean up the remainder of the WebApplicationBuilder but being more explicit about the fact that host configuration cannot be changed after builder construction. This means that the only way to change the application name, environment, and content root is via the string[] args parameter. This means we throw if the user attempts to modify the host configuration using WebApplicationBuilder.Host or WebApplicationBuilder.WebHost.UseSetting(...)
- Also used the IWebHostEnvironment provided by stashed in IHostBuilder.Properties as it is the same instance that was created by the default builder. Removed the WebHostEnvironment and tests as a result.
- Added tests

### CODE REVIEW

**Review by halter73 — APPROVED** [review]

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 1]
```diff
@@ -165,32 +165,124 @@ public void WebApplicationBuilderWebHost_ThrowsWhenBuiltDirectly()
...
+
+            var builder = WebApplication.CreateBuilder(new[] { $"--contentRoot={contentRoot}" });
+
+            builder.WebHost.UseSetting("WEBROOT", "wwwroot");
+
+            Assert.Equal("wwwroot", builder.WebHost.GetSetting("webroot"));
+
+            var app = builder.Build();
+
+            Assert.Equal(Path.Combine(contentRoot, "wwwroot"), app.Environment.WebRootPath);
```
Comment (halter73) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:283]: Do we still have any tests for a relative content root verifying it's relative to `AppContext.BaseDirectory`? It'd be similar to the first half of the `RelativePathsAreMappedToFullPaths`?

File: src/DefaultBuilder/src/BootstrapHostBuilder.cs [hunk 2]
```diff
@@ -114,26 +105,59 @@ public void RunDefaultCallbacks(HostBuilder innerBuilder)
...
             }
 
-            // Configuration doesn't auto-update during the bootstrap phase to reduce I/O,
-            // but we do need to update between host and app configuration so the right environment is used.
-            _environment.ApplyConfigurationSettings(_configuration);
+            // This is the hosting environment based on configuration we've seen so far.
+            var hostingEnvironment = new HostingEnvironment()
+            {
+                ApplicationName = _configuration[HostDefaults.ApplicationKey],
+                EnvironmentName = _configuration[HostDefaults.EnvironmentKey] ?? Environments.Production,
```
Comment (captainsafia) [inline src/DefaultBuilder/src/BootstrapHostBuilder.cs:112]: Is it OK to fallback to `Production` here?

Comment (davidfowl) [reply src/DefaultBuilder/src/BootstrapHostBuilder.cs:112]: Yes, we default to production in absence of other configuration. This logic also mimics the default host https://github.com/dotnet/runtime/blob/cf52b7e38d28b7e28358cee052effe0ed47efb2e/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L180-L196

**Review by captainsafia — APPROVED** [review]

File: src/DefaultBuilder/src/BootstrapHostBuilder.cs [hunk 3]
```diff
@@ -1,9 +1,11 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Reflection;
```
Comment (captainsafia) [inline src/DefaultBuilder/src/BootstrapHostBuilder.cs:4]: Is this used anywhere?

Comment (davidfowl) [reply src/DefaultBuilder/src/BootstrapHostBuilder.cs:4]: Yep https://github.com/dotnet/aspnetcore/pull/34829/files#diff-f0e6f715f99a336de90bdbeef327cb71b7ee811d00d7c1722049557adaa009c5R44

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs (hunk 1 above)
Comment (davidfowl) [reply src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:283]: I can add one.

Comment (davidfowl) [reply src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:283]: Will look into it.

**Pull request merged**